### PR TITLE
Optimize mypy.fastparse (Python 3 only)

### PR DESCRIPTION
**Pull request by JukkaL**
This implements various small optimizations to make AST conversion
from the typed_ast AST faster in compiled mypy. This makes compiled
mypy about 4% faster on self-check, and it also makes non-compiled
mypy around 1% faster on self-check.

Notable included optimizations:
* Implement visit methods more efficiently. Don't inherit visitors from
  `typed_ast` visitor classes.
* Refactor decorated methods away, as they are slow with mypyc.
* Refactor away many non-native attribute lookups.
* Refactor away some nested functions as they are slow with mypyc.

Also removed some dead code.

I'll implement the same optimizations for Python 2 AST conversion once
this PR has been accepted to avoid having to update two PRs based
on review.

### CODE REVIEW

File: mypy/fastparse.py [hunk 1]
```diff
@@ -179,33 +173,44 @@ def __init__(self,
...
+        node.column = n.col_offset
+        return node
 
-    def translate_expr_list(self, l: Sequence[ast3.AST]) -> List[Expression]:
+    def translate_expr_list(self, l: Sequence[AST]) -> List[Expression]:
         res = []  # type: List[Expression]
         for e in l:
             exp = self.visit(e)
-            isinstance(exp, Expression)
+            # assert isinstance(exp, Expression)
```
Comment (ilevkivskyi) [inline mypy/fastparse.py]: Can this be removed? Anyway `visit` returns `Any`.

File: mypy/fastparse.py [hunk 2]
```diff
@@ -179,33 +173,44 @@ def __init__(self,
...
             res.append(exp)
         return res
 
-    def translate_stmt_list(self, l: Sequence[ast3.AST]) -> List[Statement]:
+    def translate_stmt_list(self, l: Sequence[AST]) -> List[Statement]:
         res = []  # type: List[Statement]
         for e in l:
             stmt = self.visit(e)
-            isinstance(stmt, Statement)
+            # assert isinstance(stmt, Statement)
```
Comment (ilevkivskyi) [inline mypy/fastparse.py]: This can probably be removed as well.

File: mypy/fastparse.py [hunk 3]
```diff
@@ -324,23 +329,26 @@ def visit_Module(self, mod: ast3.Module) -> MypyFile:
...
+        return f
 
     def do_func_def(self, n: Union[ast3.FunctionDef, ast3.AsyncFunctionDef],
                     is_coroutine: bool = False) -> Union[FuncDef, Decorator]:
         """Helper shared between visit_FunctionDef and visit_AsyncFunctionDef."""
         no_type_check = bool(n.decorator_list and
                              any(is_no_type_check_decorator(d) for d in n.decorator_list))
 
-        args = self.transform_args(n.args, n.lineno, no_type_check=no_type_check)
+        lineno = n.lineno
```
Comment (ilevkivskyi) [inline mypy/fastparse.py]: Why not do the same for `n.col_offset`? (Probably a bit below when it is actually used.)

File: mypy/fastparse.py [hunk 4]
```diff
@@ -769,139 +781,140 @@ def visit_UnaryOp(self, n: ast3.UnaryOp) -> UnaryExpr:
...
-                        cast(List[Optional[str]], [None] * len(n.args)) +
-                        [k.arg for k in n.keywords])
+            [a.value if isinstance(a, Starred) else a for a in args] +
+            [k.value for k in keywords])
+        arg_kinds = ([ARG_STAR if type(a) is Starred else ARG_POS for a in args] +
+                     [ARG_STAR2 if arg is None else ARG_NAMED for arg in keyword_names])
+        e = CallExpr(self.visit(n.func),
+                     arg_types,
+                     arg_kinds,
+                     cast('List[Optional[str]]', [None] * len(args)) + keyword_names)
```
Comment (ilevkivskyi) [inline mypy/fastparse.py]: Why the quotes are needed around `List[Optional[str]]` now?

File: mypy/nodes.py [hunk 5]
```diff
@@ -742,7 +742,7 @@ def __init__(self, name: str, type: 'Optional[mypy.types.Type]' = None) -> None:
         super().__init__()
         self._name = name   # Name without module prefix
         # TODO: Should be Optional[str]
-        self._fullname = cast(Bogus[str], None)  # Name with module prefix
+        self._fullname = cast('Bogus[str]', None)  # Name with module prefix
```
Comment (ilevkivskyi) [inline mypy/nodes.py]: Why quotes are now needed here?

File: mypy/fastparse.py [hunk 6]
```diff
@@ -179,33 +173,44 @@ def __init__(self,
...
-        return super().visit(node)
+        typeobj = type(node)
+        visitor = self.visitor_cache.get(typeobj)
+        if visitor is None:
+            method = 'visit_' + node.__class__.__name__
+            visitor = getattr(self, method)
+            self.visitor_cache[typeobj] = visitor
+        return visitor(node)
+
+    def set_line(self, node: N, n: Union[ast3.expr, ast3.stmt]) -> N:
```
Comment (ilevkivskyi) [inline mypy/fastparse.py]: Just a random idea, maybe it is possible to set line at the end of `visit()` above and avoid "manually" calling it everywhere?

File: mypy/fastparse.py (hunk 3 above)
Comment (JukkaL) [reply mypy/fastparse.py]: It's only used when generating errors, so the code paths will be rare.

File: mypy/fastparse.py (hunk 4 above)
Comment (JukkaL) [reply mypy/fastparse.py]: They are not needed but they speed things up on older Python versions (when not compiled).

File: mypy/nodes.py (hunk 5 above)
Comment (JukkaL) [reply mypy/nodes.py]: Again, it's a micro-optimization. Profiling hints at spending a lot of time generating temporary type objects.

File: mypy/fastparse.py (hunk 6 above)
Comment (JukkaL) [reply mypy/fastparse.py]: It's like this because it should only be done for certain node types, not always.

**Pull request merged**